File: leetcode/tree/1325. Delete Leaves With a Given Value/py/main.py
# Time O(N)
# Space O(N)
class Solution:
    def removeLeafNodes(self, root: TreeNode, target: int) -> TreeNode:
        if not root:
          return root

        def isLeaf(node):
          return node.left == None and node.right == None

        def dfs(node):
          if node == None:
            return None

          node.left = dfs(node.left)
          node.right = dfs(node.right)

          if isLeaf(node) and node.val == target:
            return None

          return node
        
        return dfs(root)

File: leetcode/tree/1325. Delete Leaves With a Given Value/js/main.js
// Time O(N)
// Space O(N)
const removeLeafNodes = (root, target) => {
  if (root.left != null) {
    root.left = removeLeafNodes(root.left, target);
  }

  if (root.right != null) {
    root.right = removeLeafNodes(root.right, target);
  }

  return root.left == root.right && root.val == target ? null : root;
};
